src/async-components/views/dialogs/security/CreateSecretStorageDialog.tsx

```diff
@@ -31,50 +32,80 @@ import AccessibleButton from "../../../../components/views/elements/AccessibleBu
+} from '../../../../utils/WellKnownUtils';
 import SecurityCustomisations from "../../../../customisations/Security";
-
-const PHASE_LOADING = 0;
-const PHASE_LOADERROR = 1;
-const PHASE_CHOOSE_KEY_PASSPHRASE = 2;
-const PHASE_MIGRATE = 3;
-const PHASE_PASSPHRASE = 4;
-const PHASE_PASSPHRASE_CONFIRM = 5;
-const PHASE_SHOWKEY = 6;
-const PHASE_STORING = 8;
-const PHASE_CONFIRM_SKIP = 10;
+import { IValidationResult } from '../../../../components/views/elements/Validation';
+import Field from '../../../../components/views/elements/Field';
+import BaseDialog from "../../../../components/views/dialogs/BaseDialog";
+import Spinner from '../../../../components/views/elements/Spinner';
+import InteractiveAuthDialog from '../../../../components/views/dialogs/InteractiveAuthDialog';
+
+enum CreateSecretStoragePhases {
+    Loading = 0,
+    LoadError = 1,
+    ChooseKeyPassphrase = 2,
+    Migrate = 3,
+    Passphrase = 4,
+    PassphraseConfirm = 5,
+    ShowKey = 6,
+    Storing = 8,
+    ConfirmSkip = 10,
+}
 
 const PASSWORD_MIN_SCORE = 4; // So secure, many characters, much complex, wow, etc, etc.
 
-// these end up as strings from being values in the radio buttons, so just use strings
-const CREATE_STORAGE_OPTION_KEY = 'key';
-const CREATE_STORAGE_OPTION_PASSPHRASE = 'passphrase';
+interface IProps {
+    hasCancel?: boolean;
+    accountPassword?: string;
+    forceReset?: boolean;
+    onFinished: (success: boolean) => void;
+}
+
+interface IState {
+    phase: CreateSecretStoragePhases;
+    passPhrase: string;
+    passPhraseValid: boolean;
+    passPhraseConfirm: string;
+    passPhraseKeySelected?: SecureBackupSetupMethod;
+    copied: boolean;
+    downloaded: boolean;
+    setPassphrase: boolean;
+    backupInfo: IKeyBackupInfo;
+    backupSigStatus: TrustInfo;
+    // does the server offer a UI auth flow with just m.login.password
+    // for /keys/device_signing/upload?
+    canUploadKeysWithPasswordOnly: boolean;
+    accountPassword: string;
+    accountPasswordCorrect: boolean;
+    canSkip: boolean;
+    error?: string | null;
+}
 
 /*
  * Walks the user through the process of creating a passphrase to guard Secure
  * Secret Storage in account data.
  */
-export default class CreateSecretStorageDialog extends React.PureComponent {
-    static propTypes = {
-        hasCancel: PropTypes.bool,
-        accountPassword: PropTypes.string,
-        forceReset: PropTypes.bool,
-    };
-
+export default class CreateSecretStorageDialog extends React.PureComponent<IProps, IState> {
     static defaultProps = {
         hasCancel: true,
         forceReset: false,
     };
 
-    constructor(props) {
-        super(props);
+    private recoveryKey: IRecoveryKey = null;
+    private recoveryKeyNode: HTMLElement = null;
+    private backupKey: Uint8Array = null;
 
-        this._recoveryKey = null;
-        this._recoveryKeyNode = null;
-        this._backupKey = null;
+    private passphraseField: React.RefObject<Field> = createRef();
```

Comment:
```suggestion
    private passphraseField = createRef<Field>();
```

cleaner and shorter IMO